tests/integration_tests/functional/test_vsock.py

```diff
@@ -225,54 +225,64 @@ def test_vsock_transport_reset_g2h(uvm_nano, microvm_factory):
-    time.sleep(0.5)
-    assert Path(host_socket_path).exists()
-    test_vm.create_jailed_resource(host_socket_path)
-
-    # Create a socat process in the guest which will connect to the host socat
-    guest_socat_commmand = f"tmux new -d 'socat - vsock-connect:2:{ECHO_SERVER_PORT}'"
-    test_vm.ssh.run(guest_socat_commmand)
-
-    # socat should be running in the guest now
-    code, _, _ = test_vm.ssh.run("pidof socat")
-    assert code == 0
-
     # Create snapshot.
     snapshot = test_vm.snapshot_full()
     test_vm.resume()
 
-    # After `create_snapshot` + 'restore' calls, connection should be dropped
-    code, _, _ = test_vm.ssh.run("pidof socat")
-    assert code == 1
-
-    # Kill host socat as it is not useful anymore
-    host_socat.kill()
-    host_socat.communicate()
-
     # Terminate VM.
     test_vm.kill()
 
-    # Load snapshot.
-    vm2 = microvm_factory.build()
-    vm2.spawn()
-    vm2.restore_from_snapshot(snapshot, resume=True)
-
-    # After snap restore all vsock connections should be
-    # dropped. This means guest socat should exit same way
-    # as it did after snapshot was taken.
-    code, _, _ = vm2.ssh.run("pidof socat")
-    assert code == 1
+    for _ in range(5):
+        # Load snapshot.
+        new_vm = microvm_factory.build()
+        new_vm.spawn()
+        new_vm.restore_from_snapshot(snapshot, resume=True)
+
+        # After snap restore all vsock connections should be
+        # dropped. This means guest socat should exit same way
+        # as it did after snapshot was taken.
+        code, _, _ = new_vm.ssh.run("pidof socat")
+        assert code == 1
+
+        host_socket_path = os.path.join(
+            new_vm.path, f"{VSOCK_UDS_PATH}_{ECHO_SERVER_PORT}"
+        )
+        host_socat_commmand = [
+            "socat",
+            "-dddd",
+            f"UNIX-LISTEN:{host_socket_path},fork",
+            "STDOUT",
+        ]
+        host_socat = subprocess.Popen(
+            host_socat_commmand, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+
+        # Give some time for host socat to create socket
+        time.sleep(0.5)
+        assert Path(host_socket_path).exists()
+        new_vm.create_jailed_resource(host_socket_path)
+
+        # Create a socat process in the guest which will connect to the host socat
+        guest_socat_commmand = (
+            f"tmux new -d 'socat - vsock-connect:2:{ECHO_SERVER_PORT}'"
+        )
+        new_vm.ssh.run(guest_socat_commmand)
+
+        # socat should be running in the guest now
+        code, _, _ = new_vm.ssh.run("pidof socat")
+        assert code == 0
+
+        # Create snapshot.
+        snapshot = new_vm.snapshot_full()
+        new_vm.resume()
+
+        # After `create_snapshot` + 'restore' calls, connection should be dropped
+        code, _, _ = new_vm.ssh.run("pidof socat")
+        assert code == 1
+
+        # Kill host socat as it is not useful anymore
+        host_socat.kill()
+        host_socat.communicate()
```

Comment:
we could re-use the socat echo server between runs, but for that we would need a path outside the VM root.

Reply:
Don't think there will be meaningful difference compared to added complexity.